Emit DW_AT_main_subprogram

tromey opened the pull request:
This changes rustc to emit DW_AT_main_subprogram on the "main" program.
This lets gdb suitably stop at the user's main in response to
"start" (rather than the library's main, which is what happens
currently).

Fixes #32620
r? michaelwoerister

alexcrichton commented:
It looks like llvm 3.7 may break the tests added here (on travis), but I think that can be fixed with `// min-llvm-version 3.7`

tromey commented:
> I think that can be fixed with // min-llvm-version 3.7

Thanks - I will update the patch soon.  I think for this I'll introduce a new test, so that the existing test keeps running on travis.

michaelwoerister commented on src/librustc_llvm/ffi.rs:
```diff
@@ -451,6 +451,7 @@ pub mod debuginfo {
         FlagIndirectVariable = 1 << 13,
         FlagLValueReference = 1 << 14,
         FlagRValueReference = 1 << 15,
+        FlagMainSubprogram = 1 << 21,
```
I looks like older LLVM versions will just silently ignore this. At least 3.7 on Travis. Which is good.

tromey commented on src/librustc_trans/debuginfo/mod.rs:
```diff
@@ -286,7 +296,7 @@ pub fn create_function_debug_context<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
             true,
             scope_line as c_uint,
-            FlagPrototyped as c_uint,
+            flags,
```
`flags` is initialized to include it:

```
+    let mut flags = FlagPrototyped as c_uint;
```

I'm happy to write this some other way if you think it would be clearer.

michaelwoerister replied:
Oh, that's right, I just missed it. Sorry.

tromey commented:
> It looks like llvm 3.7 may break the tests added here (on travis), but I think that can be fixed with // min-llvm-version 3.7

I'm wondering today if that should be "min-llvm-version 3.8"?  Since IIUC 3.7 is what is in travis, and will fail.  Even 3.8 is technically wrong in a sense, because if there's a real 3.8 release, it won't contain the needed patch.

tromey commented on src/librustc_trans/debuginfo/mod.rs:
```diff
@@ -274,6 +274,16 @@ pub fn create_function_debug_context<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
+    let mut flags = FlagPrototyped as c_uint;
+    match *cx.sess().entry_fn.borrow() {
+        Some((id, _)) => {
+            if instance == Instance::mono(cx.shared(), cx.tcx().map.local_def_id(id)) {
```
I needed something a bit wordier:

```
            if let Some(local_id_unwrapped) = local_id {
                if id == local_id_unwrapped {
                    flags |= FlagMainSubprogram as c_uint;
                }
            }
```

michaelwoerister replied:
Oh, that makes sense. Though you could probably make it shorter be wrapping `id` in an `Option` and then comparing `Option` against `Option`. Or do something like this:
```rust
let entry_fn_id = *cx.sess().entry_fn.borrow();
if entry_fn_id.is_some() && local_id == entry_fn_id {
    flags |= ...;
}
```

arielb1 replied:
The standard style is is `if entry_fn_id == Some(local_id)`.

michaelwoerister commented:
> I'm wondering today if that should be "min-llvm-version 3.8"? 

I guess that no LLVM release in existence today contains the patch, so this test can only pass with our own version of LLVM. Do we have a `rust-llvm-only` directive for tests?

bors commented:
:umbrella: The latest upstream changes (presumably #37857) made this pull request unmergeable. Please resolve the merge conflicts.

tromey commented:
Just FYI - I've rebased this and fixed it up, and it works in manual testing.  However the test cases fail now.  I don't know the test suite setup very well; my suspicion is that the tests aren't built in isolation, leading rustc not to emit `DW_AT_main_subprogram` for the test `main` functions.  I'll try to ping someone on irc tomorrow to see what can be done; or else if you are reading this and happen to know, please comment.  Thanks :)

tromey commented:
> However the test cases fail now.

Of course this turned out to be operator error; somehow I had a stale rustc in my tree that I was running manually.

michaelwoerister commented:
Let me know when it's ready for review.

tromey commented:
@michaelwoerister - hopefully it's ready now.

tromey commented:
Well, I guess not.  In the new scheme for DIFlags I suppose some kind of version or other configury is needed.  Maybe I should have listened to that warning in RustWrapper.cpp.

bors commented:
:umbrella: The latest upstream changes (presumably #38701) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
ping @tromey, it looks like this may need a rebase now?

tromey commented:
> it looks like this may need a rebase now?

Yeah.  Sorry about my silence on this.  I still intend to work on it.  I think there are two issues with the patch currently.

One is that the refactor in this area means that the old approach of passing in the constant, regardless of the underlying LLVM version, will no longer work, because the enumerator in question just doesn't exist in earlier versions of LLVM.  The Rust fork of LLVM does have this, and so does 4.0, but I don't know of a way to distinguish the Rust LLVM from the ordinary upstream LLVM.  So, maybe adding a way to distinguish this case would be good.

Alternatively a simpler solution would be to only enable this feature for LLVM 4.0.  I think this is less nice because in the future I hope to add more Rust-specific features to LLVM's debuginfo generation, and backport these to the Rust LLVM fork; and if this all comes to pass then it will be handy to have the necessary infrastructure already in place.

The second problem with this patch is related: the tests also have no way to check whether they're being run against the Rust fork of LLVM.

cuviper commented:
> I don't know of a way to distinguish the Rust LLVM from the ordinary upstream LLVM. So, maybe adding a way to distinguish this case would be good.

There is a `-DLLVM_RUSTLLVM` for this sort of thing.  It's used in `PassWrapper.cpp` to stub a couple functions that only exist in the Rust fork of LLVM.

tromey commented:
> There is a -DLLVM_RUSTLLVM for this sort of thing

Thanks!  That really simplifies things.

alexcrichton commented:
re-r? @michaelwoerister

michaelwoerister commented:
@bors r+

Thanks @tromey!

bors commented:
:pushpin: Commit b037c52 has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister b037c5211b9ba496252af5a0369b48649f91ccc8 -->

bors commented:
:hourglass: Testing commit b037c52 with merge 4053276...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1877), [status-travis](https://travis-ci.org/rust-lang/rust/builds/200053624)
Approved by: michaelwoerister
Pushing 405327635419e22a956dfe8f7caf4817c8ae5e93 to master...
